scripts/ci/build_emscripten.sh

```diff
@@ -34,55 +34,106 @@
-# also used with earlier clang versions.
-# TODO: This can be removed if and when all usages of `move()` in our codebase use the `std::` qualifier.
-CMAKE_CXX_FLAGS="-Wno-unqualified-std-cast-call"
+    # Disable warnings for unqualified `move()` calls, introduced and enabled by
+    # default in clang-16 which is what the emscripten docker image uses.
+    # Additionally, disable the warning for unknown warnings here, as this script is
+    # also used with earlier clang versions.
+    # TODO: This can be removed if and when all usages of `move()` in our codebase use the `std::` qualifier.
+    CMAKE_CXX_FLAGS="-Wno-unqualified-std-cast-call"
 
+    mkdir -p "$build_dir"
+    cd "$build_dir"
+    emcmake cmake \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DBoost_USE_STATIC_LIBS=1 \
+        -DBoost_USE_STATIC_RUNTIME=1 \
+        -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS}" \
+        -DTESTS=0 \
+    ..
+    make soljson
 
-mkdir -p "$BUILD_DIR"
-cd "$BUILD_DIR"
-emcmake cmake \
-	-DCMAKE_BUILD_TYPE=Release \
-	-DBoost_USE_STATIC_LIBS=1 \
-	-DBoost_USE_STATIC_RUNTIME=1 \
-	-DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS}" \
-	-DTESTS=0 \
-  ..
-make soljson
+    cd ..
+    mkdir -p upload
+    scripts/ci/pack_soljson.sh "$build_dir/libsolc/soljson.js" "$build_dir/libsolc/soljson.wasm" upload/soljson.js
+    cp upload/soljson.js ./
 
-cd ..
-mkdir -p upload
-scripts/ci/pack_soljson.sh "$BUILD_DIR/libsolc/soljson.js" "$BUILD_DIR/libsolc/soljson.wasm" upload/soljson.js
-cp upload/soljson.js ./
+    OUTPUT_SIZE=$(ls -la soljson.js)
 
-OUTPUT_SIZE=$(ls -la soljson.js)
+    echo "Emscripten output size: $OUTPUT_SIZE"
+}
 
-echo "Emscripten output size: $OUTPUT_SIZE"
+function show_help() {
+cat << EOF
+Usage: ${0##*/} [-h|--help] [--build-dir DIR] [--prerelease-source prerelease_source]
+Build Solidity emscripten binary
+    -h | --help  Display this help message
+    --build-dir  The emscripten build directory
+    --prerelease-source The prerelease source string. E.g. 'nightly' or 'ci'.
+EOF
+}
+
+function main() {
+    local build_dir="emscripten_build"
+    local prerelease_source=""
+
+    while :; do
+        case "$1" in
+            -h|--help)
+                show_help
+                exit 0
+                ;;
+            --build-dir)
+                if [[ -n "$2" ]]; then
+                    build_dir="$2"
+                    shift
+                else
+                    fail "Option --build-dir cannot be empty"
+                fi
+                shift
+                ;;
+            --prerelease-source)
+                if [[ -n "$2" ]]; then
+                    prerelease_source="$2"
+                    shift
+                else
+                    fail "Option --prerelease-source cannot be empty"
+                fi
+                shift
+                ;;
+            *)
+                if [[ -n "$1" ]]; then
+                    fail "Invalid option: $1"
+                fi
+                break
```

Comment:
If I understand this correctly - this will ignore empty arguments (something like `build_emscripten.sh ""`)? I think it would be fine (and simpler) to just always report an error:
```suggestion
            *) fail "Invalid option: $1"
```

Reply:
No, the default case will also be called if you don't provide any parameters or provide invalid parameters. The `if` is to get the latter. The first is for compatibility reasons since the script can run without parameters and use the default values set in the main function. Just throwing a fail in the default case prevents the script from running without parameters.

Reply:
But I guess that's because you're using an infinite loop? Just realized that `while :` is basically `while true`. Why not instead check if there are any parameters left with `$#`?

Reply:
Yes. It will also work, but it will require the break in case the parameter is empty anyway.
Like:
```bash
    while [[ $# -gt 0 ]] ; do
        case "$1" in
            #...
            *) [[ -z "$1" ]] || fail "Invalid option: $1"
               break ;;
        esac
    done
```
If you don't do the `[[ -z "$1" ]]` and only call `fail` like below:
```bash
*) fail "Invalid option: $1"
```
If the script is called without parameters (i.e. `./build_emscripten.sh`) it would fail with `invalid option` error as well, instead of using the default values.

Reply:
That's strange. I just tried it on a dummy script and it does not behave that way for me. Try this:
```bash
while [[ $# -gt 0 ]] ; do
    case "$1" in
        x) echo x ; shift ;;
        *) echo y ; exit 1 ;;
    esac
done
```

When I run it without parameters, it does not print anything so it does not even enter the default branch.

Reply:
Hum...yes, you are right. It turns out that the issue was because we added this here https://github.com/ethereum/solidity/pull/13581#discussion_r1007958515 to escape whitespace on paths. So when calling the script without argument, the string "`'' `"  is given as an argument to the function (note the double `'` and space added by the printf `"%q "`), thus entering the while loop, which does not happen in your example.
Your solution indeed works as expected when passing the argument directly (without the `printf`). I managed to reproduce it here based on your code snippet:

```bash
#!/bin/bash

function fail() {
    echo "fail"
    exit 1
}

function f1() {
    echo "f1 params: $* | $#"
    param="default"
    while [[ $# -gt 0 ]] ; do
        case "$1" in
            x) param="x" ; shift ;;
            *) [[ -z "$1" || "$1" == "'' " ]] || fail
                break ;;
        esac
    done
    echo "Got: $param"
}

function f2() {
    echo "f2 params: $* | $#"
    param="default"
    while [[ $# -gt 0 ]] ; do
        case "$1" in
            x) param="x" ; shift ;;
            *) fail ;;
        esac
    done
    echo "Got: $param"
}

f1 "$@"
f2 "$@"

f1 "$(printf "%q " "${@}")"
f2 "$(printf "%q " "${@}")"
```

Reply:
Forget about it. It is fixed now using `?*` to handle incorrect parameters.

Reply:
Ok. I see. Nice that we finally know what's happening here.

I think that the current approach of failing only on some specific patterns is not great through. We should do the opposite - handle known inputs, error out on anything else. Also, `'' ` is simply not valid input for `scripts.ci/build_emscripten.sh` it should not have to deal with that. If we're getting that weird input, we should deal with it in `scripts/build_emscripten.sh` instead.

Can you just add an `if` in `scripts/build_emscripten.sh` to use empty input if it reduces to `'' `?